File: leetcode/p00771/p00771.go
package p00771

// Time Complexity: O(n)
// Space Complexity: O(n)
func numJewelsInStones(jewels string, stones string) int {
	mapping := map[rune]bool{}

	for _, ch := range jewels {
		mapping[ch] = true
	}

	count := 0
	for _, ch := range stones {
		if _, exists := mapping[ch]; exists {
			count++
		}
	}

	return count
}

File: leetcode/p00771/p00771.py
class Solution:
    # Time Complexity: O(n)
    # Space Complexity: O(n)
    def numJewelsInStones(self, jewels: str, stones: str) -> int:
        mapping = set(jewels)

        count = 0
        for ch in stones:
            if ch in mapping:
                count += 1

        return count
    
if __name__ == '__main__':
    s = Solution()

    assert s.numJewelsInStones("aA", "aAAbbbb") == 3
    assert s.numJewelsInStones("z", "ZZ") == 0